### PR TITLE
Move SSA conversion earlier in the analysis phase

### PR DESCRIPTION
**Pull request by ptomin**
- `Architectures`:
    - Return address on stack
    - Get register by domain and bit range
    - `GrfToString`/`GetFlagGroup`
    - `GetSubFlags`
    - X86:
        - explicit FPU stack
- `Core`:
    - Refactor `ApplicationBuilder `, create `FrameApplicationBuilder`, `CallApplicationBuilder` subclasses
    - Introduce `BitRange`
    - Move `CallRewriter` to `Decompiler.Analysis`
    - Introduce `CallBinding` to store uses and definitions of `CallInstruction`
    - Use(Identifier) -> Use(Expression)
    - Modify `ApplyConstants` in `Operator` subclasses
    - Use `TypeReferenceFormatter` instead of `TypeFormatter`
    - `SequenceStorage` generalized to support more than two sub-storage
    - New methods were added to `Storage`: `OverlapsWith`, `Covers`, `Exceeds`, `GetBitRange`.
    - Add `IsPointer` and `IsIntegral` properties to `DataType`
- `Decompiler.Analysis`:
    - Remove `AliasState` and `Aliases`
    - Remove  `GlobalCallRewriter`
    - New `SSA` transform algorithm
    - New `TrashedRigisterFinder` for recursive functions
    - Create `UnusedOutValuesRemover`/`UserRegisterFinder`
    - Move inserting of initialisation statements for user defined parameters from `UserSignatureBuilder` to `SSA` transformation
    - `WebBuilder`: do not insert declarations for unused identifiers
- `Decompiler.Evaluation`:
   - Add new simplification rules to `ExpressionSimplifiler`

### CODE REVIEW

File: src/Decompiler/Analysis/SsaTransform.cs [hunk 1]
```diff
@@ -27,164 +27,893 @@
...
+                }
+                //$REVIEW: this is very x86/x87 specific; find a way to generalize
+                // this to any sort of stack-based discipline.
+                foreach (var def in calleeFlow.Trashed.OfType<FpuStackStorage>()
+                    .Where(def => def.FpuStackOffset >= fpuStackDelta))
+                {
+                    var fpuDefExpr = arch.CreateFpuStackAccess(
+                        ssa.Procedure.Frame, 
+                        def.FpuStackOffset,
+                        PrimitiveType.Word64); //$TODO: datatype?
```
Comment (ptomin) [inline src/Decompiler/Analysis/SsaTransform.cs]: @uxmal This code breaks assumption that all definitions of `CallInstruction` are `Identifiers` 
So code like that https://github.com/uxmal/reko/blob/d97a894bef758e15beba2f4e8e2af629e97d9676/src/Decompiler/Analysis/TrashedRegisterFinder3.cs#L403 will cause cast exception.
Possible solution is 0a1a91f: pass temporary ids to definitions of `CallInstruction` and then store them to corresponding FPU stack offsets

Comment (uxmal) [reply src/Decompiler/Analysis/SsaTransform.cs]: How will this affect live-out computations later? Perhaps you could test this by making your suggested change. That is, have the `call` instruction write its outputs into temp variables. Then, see how it affects the liveness and trashed register analyses.

Comment (ptomin) [reply src/Decompiler/Analysis/SsaTransform.cs]: All tests (except 3) passed. New output for that failed tests you can see in 0a1a91f.
And it gives appropriate output on `Blade.dcproject`

Comment (ptomin) [reply src/Decompiler/Analysis/SsaTransform.cs]: Also I have changed `UsedRegisterFinder` in 0a1a91f  to make live-out computation work correctly

Comment (ptomin) [reply src/Decompiler/Analysis/SsaTransform.cs]: I'll raise PR(s) this weekend

File: subjects/Elf/RiscV/ipcalc/ipcalc.c [hunk 2]
```diff
@@ -4,8 +4,6 @@
 
 #include "ipcalc.h"
 
-//$BUG: if we can prove that these are .PLT stubs, remove them 
-// from the output.
```
Comment (ptomin) [inline subjects/Elf/RiscV/ipcalc/ipcalc.c]: $BUG: if we can prove that these are .PLT stubs, remove them 
from the output.

File: subjects/Elf/RiscV/ipcalc/ipcalc.c [hunk 3]
```diff
@@ -856,7 +854,6 @@ word64 fn0000000000015FCC(struct Eq_3 * gp, Eq_529 a0, ptr64 & a2Out, ptr64 & a3
 	FILE * a3_18 = (word64) gp->qwFFFFF880;
 	if (a0 != 0x00)
 	{
-        //$BUG: all these DPB....
```
Comment (ptomin) [inline subjects/Elf/RiscV/ipcalc/ipcalc.c]: $BUG: all these DPB....

File: subjects/Elf/RiscV/ipcalc/ipcalc.c [hunk 4]
```diff
@@ -891,7 +888,6 @@ word64 fn0000000000015FCC(struct Eq_3 * gp, Eq_529 a0, ptr64 & a2Out, ptr64 & a3
 		a0_350 = DPB(0x00011348, fwrite(&globals->v11348, 0x01, 0x3A, (word64) gp->qwFFFFF880), 0);
 		a0_361 = DPB(0x00011388, fwrite(&globals->v11388, 0x01, 0x39, (word64) gp->qwFFFFF880), 0);
 		a0_375 = DPB(70600, fwrite(&globals->v113C8, 0x01, 0x3E, (word64) gp->qwFFFFF880), 0);
-//$BUG: why are a2 and a3 live out here?
```
Comment (ptomin) [inline subjects/Elf/RiscV/ipcalc/ipcalc.c]: $BUG: why are a2 and a3 live out here?

File: subjects/Elf/Sparc/rtems/sparc-rtems-unprotoize.c [hunk 5]
```diff
@@ -111,7 +111,6 @@ Eq_161 xmalloc(Eq_161 o0, ptr32 & i1Out, ptr32 & i2Out, ptr32 & i6Out, ptr32 & i
 {
 	if (malloc(o0) != null)
 	{
-        //$BUG: we're not returning the result of malloc
```
Comment (ptomin) [inline subjects/Elf/Sparc/rtems/sparc-rtems-unprotoize.c]: $BUG: we're not returning the result of malloc

File: subjects/Elf/Sparc/rtems/sparc-rtems-unprotoize.c [hunk 6]
```diff
@@ -135,7 +134,6 @@ Eq_161 xmalloc(Eq_161 o0, ptr32 & i1Out, ptr32 & i2Out, ptr32 & i6Out, ptr32 & i
 		word32 stackArg80 = <invalid>;
 		word32 stackArg84 = <invalid>;
 		word32 stackArg88 = <invalid>;
-		//$BUG: 0x000170A8 is a pointer to a string; show a string constant
```
Comment (ptomin) [inline subjects/Elf/Sparc/rtems/sparc-rtems-unprotoize.c]: $BUG: 0x000170A8 is a pointer to a string; show a string constant

File: subjects/Elf/Sparc/rtems/sparc-rtems-unprotoize.c [hunk 7]
```diff
@@ -145,7 +143,6 @@ Eq_161 xmalloc(Eq_161 o0, ptr32 & i1Out, ptr32 & i2Out, ptr32 & i6Out, ptr32 & i
 Eq_161 xrealloc(Eq_161 o0, Eq_161 o1, union Eq_161 & i1Out, ptr32 & i2Out)
 {
 	Eq_227 Z_34;
-    //$BUG: cond should have been replaced with a condition.
```
Comment (ptomin) [inline subjects/Elf/Sparc/rtems/sparc-rtems-unprotoize.c]: $BUG: cond should have been replaced with a condition.

File: subjects/Elf/Sparc/rtems/sparc-rtems-unprotoize.c [hunk 8]
```diff
@@ -310,7 +307,6 @@ Eq_161 safe_read(Eq_161 o0, Eq_407 o1, Eq_161 o2, struct stat & l1Out, union Eq_
 l00011864_ds_f:
 	branch l0_21 != 0x00 l00011874_ds_t
 l00011864_ds_t:
-    //$BUG: cond should be eliminated.
```
Comment (ptomin) [inline subjects/Elf/Sparc/rtems/sparc-rtems-unprotoize.c]: //$BUG: cond should be eliminated.

File: subjects/Elf/Sparc/rtems/sparc-rtems-unprotoize.c [hunk 9]
```diff
@@ -617,7 +613,6 @@ struct Eq_799 * visit_each_hash_node(struct Eq_799 * o0)
 // 00011C3C: Register Eq_832 add_symbol(Register Eq_832 o0, Register Eq_161 o1, Register out ptr32 i1Out)
 Eq_832 add_symbol(Eq_832 o0, Eq_161 o1, ptr32 & i1Out)
 {
-    //$BUG: this is a regression.
```
Comment (ptomin) [inline subjects/Elf/Sparc/rtems/sparc-rtems-unprotoize.c]: $BUG: this is a regression.

File: subjects/Elf/Sparc/rtems/sparc-rtems-unprotoize.c [hunk 10]
```diff
@@ -1116,7 +1111,6 @@ Eq_161 find_file(Eq_161 o0, char & l2Out, ptr32 & l3Out, union Eq_161 & i0Out, p
 	}
 	Eq_161 l4_101;
 	struct Eq_1719 * i6_134;
-   //$BUG: throwing away the result of malloc.
```
Comment (ptomin) [inline subjects/Elf/Sparc/rtems/sparc-rtems-unprotoize.c]: $BUG: throwing away the result of malloc.

File: subjects/Elf/x86-64/ais3_crackme/ais3_crackme.c [hunk 11]
```diff
@@ -4,7 +4,6 @@
 
 #include "ais3_crackme.h"
 
-//$BUG: _init() surely doesn't return anything?
```
Comment (ptomin) [inline subjects/Elf/x86-64/ais3_crackme/ais3_crackme.c]: $BUG: _init() surely doesn't return anything?

File: subjects/PE/x86/import-ordinals/main.c [hunk 12]
```diff
@@ -414,7 +414,6 @@ Eq_959 fn0040188B(HMODULE & ecxOut)
 	if (eax_12 != null)
 	{
 		eax_12 = (struct HINSTANCE__ *) 23117;
-        //$BUG: why unused?
```
Comment (ptomin) [inline subjects/PE/x86/import-ordinals/main.c]: $BUG: why unused?

File: subjects/regressions/snowman-51/switch.c [hunk 13]
```diff
@@ -22,8 +22,6 @@ char * get(uint32 n)
 	}
 }
 
-//$BUG: why did we lose the Win32 types here?
-
 // 10071080: Register Eq_13 DllMain(Stack Eq_14 hModule, Stack Eq_15 dwReason, Stack Eq_16 lpReserved)
 Eq_13 DllMain(Eq_14 hModule, Eq_15 dwReason, Eq_16 lpReserved)
```
Comment (ptomin) [inline subjects/regressions/snowman-51/switch.c]: $BUG: why did we lose the Win32 types here?

File: subjects/regressions/snowman-74/int16.c [hunk 14]
```diff
@@ -10,7 +10,6 @@ int32 A::get(int32 dwArg04)
 	return this->dw0000 + dwArg04;
 }
 
-//$BUG: lost the Win32 type names
 // 10071010: Register Eq_9 DllMain(Stack Eq_10 hModule, Stack Eq_11 dwReason, Stack Eq_12 lpReserved)
 Eq_9 DllMain(Eq_10 hModule, Eq_11 dwReason, Eq_12 lpReserved)
```
Comment (ptomin) [inline subjects/regressions/snowman-74/int16.c]: $BUG: lost the Win32 type names

File: subjects/PE/x86/VCExeSample/VCExeSample.c [hunk 15]
```diff
@@ -43,71 +43,68 @@ void test5()
 	globals->gbl_c->vtbl->method04(globals->gbl_c, 999, globals->r4020EC);
 }
 
-// 00401120: void test6(Stack Eq_62 c, Stack int32 a, Stack int32 b)
-void test6(Eq_62 c, int32 a, int32 b)
+// 00401120: void test6(Stack Eq_55 c, Stack int32 a, Stack int32 b)
+void test6(Eq_55 c, int32 a, int32 b)
 {
-	c->vtbl->method04(c, c->vtbl->sum(c, a, b));
+	c.u0->vtbl->method04(c, c.u0->vtbl->sum(c, a, b));
```
Comment (ptomin) [inline subjects/PE/x86/VCExeSample/VCExeSample.c]: @uxmal This looks like regression. I'll try to find the reason of such change

File: subjects/PE/x86-64/varargs_test/varargs_test.c [hunk 16]
```diff
@@ -4,13 +4,47 @@
...
+	// Please report this issue at https://github.com/uxmal/reko
+	// Failed to bind call argument.
+	// Please report this issue at https://github.com/uxmal/reko
+	ui64 rax_n = globals->t40003000 ^ fp - 200;
+	Eq_n xmm0_n = globals->t40002260;
+	Eq_n stackArg8 = <invalid>;
+	Eq_n stackArg16 = <invalid>;
+	Eq_n stackArg24 = <invalid>;
+	Eq_n stackArg32 = <invalid>;
+	fn0000000140001140(rax_n, 0x140002210, 0x03, rbx, rsi, rdi, 0x140002250, 0x140002240, stackArg8, stackArg16, stackArg24, stackArg32);
```
Comment (ptomin) [inline subjects/PE/x86-64/varargs_test/varargs_test.c]: @uxmal Looks like regression, is not it?

File: src/Core/Storage.cs [hunk 17]
```diff
@@ -158,13 +301,24 @@ public override T Accept<T>(StorageVisitor<T> visitor)
             return visitor.VisitFpuStackStorage(this, context);
         }
 
+        public override bool Covers(Storage that)
+        {
+            return that is FpuStackStorage &&
+                this.Number == that.Number;
```
Comment (ptomin) [inline src/Core/Storage.cs]: @uxmal `Number` seems to be never set, looks like incorrect code

File: src/Core/StorageBinder.cs [hunk 18]
```diff
@@ -119,46 +115,40 @@ public Identifier EnsureRegister(RegisterStorage reg)
...
             return id;
         }
 
-        public Identifier EnsureSequence(Storage head, Storage tail, DataType dataType)
+        public Identifier EnsureSequence(DataType dataType, params Storage [] elements)
         {
-            Identifier id;
-            Dictionary<Storage, Identifier> seqs;
-            if (!this.seqs.TryGetValue(head, out seqs))
+            var stg = new SequenceStorage(elements);
```
Comment (ptomin) [inline src/Core/StorageBinder.cs:128]: @uxmal Unused variable?

File: src/Core/StorageBinder.cs [hunk 19]
```diff
@@ -119,46 +115,40 @@ public Identifier EnsureRegister(RegisterStorage reg)
...
             ids.Add(id);
             return id;
         }
 
 
-        public Identifier EnsureSequence(string name, Storage head, Storage tail, DataType dataType)
+        public Identifier EnsureSequence(DataType dataType, string name, params Storage [] elements)
         {
-            if (!this.seqs.TryGetValue(head, out var seqs))
+            var stg = new SequenceStorage(elements);
```
Comment (ptomin) [inline src/Core/StorageBinder.cs:144]: @uxmal Unused variable?

**Pull request closed without merging**